pkg/gather/ssh/agent.go

```diff
@@ -0,0 +1,36 @@
+package ssh
+
+import (
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/ssh/agent"
+
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+)
+
+// newAgent initializes an SSH Agent with the keys.
+// If not keys are provided, it loads all the keys from the user's environment.
+func newAgent(keyPaths []string) (agent.Agent, error) {
+	keys, err := loadKeys(keyPaths)
+	if err != nil {
+		return nil, err
+	}
+
+	ag := agent.NewKeyring()
+	var errs []error
+	for idx := range keys {
+		if err := ag.Add(agent.AddedKey{PrivateKey: keys[idx]}); err != nil {
+			errs = append(errs, errors.Wrap(err, "failed to add key to agent"))
+		}
+	}
+	if agg := utilerrors.NewAggregate(errs); agg != nil {
+		return nil, agg
```

Comment:
No harm in returning `ag` alongside an error is there?  Then we can simplify to:

```go
return ag, utilerrors.NewAggregate(errs)
```

because [the `NewAggregate` docs][1] say:

>  If the slice is empty, this returns nil.

[1]: https://godoc.org/k8s.io/apimachinery/pkg/util/errors#NewAggregate

Reply:
your suggestion would allow the consumer capable of using an agent that has only part of the keys added to it, which I do not want.

Reply:
> your suggestion would allow the consumer capable of using an agent that has only part of the keys added to it, which I do not want.

If they decide to ignore the returned error.  Folks who ignore errors will sometimes break in unexpected (to them, because they didn't check the error) ways.  I don't think it's on us to protect them further.  But :shrug: your way is only a few lines ;).

Reply:
> > your suggestion would allow the consumer capable of using an agent that has only part of the keys added to it, which I do not want.
> 
> If they decide to ignore the returned error. Folks who ignore errors will sometimes break in unexpected (to them, because they didn't check the error) ways. I don't think it's on us to protect them further. But :shrug: your way is only a few lines ;).

Well this is not about the `only a few lines`; But more about the API of the function. Your suggestion makes `partially populated agent` object part of the API when import of the `some` of the keys failed. While the current setup makes it all or none.

Reply:
I'm fine with an API that is "here's as far as I got, and the error that made me decide to stop".  It gives the caller the most flexibility.  In this particular case, I'm fine returning nothing or the partial agent on error as the API, and partial agent has the additional benefit of simplifying the code slightly.  But I'm fine either way ;)